PLT-7583 Validator optimizations

bwbush opened the pull request:
1. Use `PlutusTx.asData` for `State`.
2. Use `PlutusTx.asData` for `Action`.
3. Faster `Value.geq`, but with fallback.
4. Tracing and profiling, including flamegraphs.
5. Severed dependency of validators on `marlowe-cardano` repository, but tests still rely on that (for consistency).

***Note to reviewers***

Here is an efficient way to compare the validator files where were renamed or relocated:

```bash
git clone git@github.com:input-output-hk/marlowe-plutus --single-branch \
  --branch PLT-7583 marlowe-plutus.new
git clone git@github.com:input-output-hk/marlowe-plutus --single-branch \
  --branch main marlowe-plutus.main
git clone git@github.com:input-output-hk/marlowe-cardano --single-branch \
  --branch PLT-8148-plutus1.15.0.0-feature marlowe-cardano.feature
vimdiff marlowe-plutus.new/marlowe-plutus/src/Language/Marlowe/Plutus/Semantics/Types/Address.hs \
        marlowe-cardano.feature/marlowe/src/Language/Marlowe/Core/V1/Semantics/Types/Address.hs 
vimdiff marlowe-plutus.new/marlowe-plutus/src/Language/Marlowe/Plutus/Semantics/Types.hs \
        marlowe-cardano.feature/marlowe/src/Language/Marlowe/Core/V1/Semantics/Types.hs 
vimdiff marlowe-plutus.new/marlowe-plutus/src/Language/Marlowe/Plutus/Semantics.hs \
        marlowe-cardano.feature/marlowe/src/Language/Marlowe/Core/V1/Semantics.hs 
vimdiff marlowe-plutus.new/marlowe-plutus/src/Language/Marlowe/Plutus/ScriptTypes.hs \
        marlowe-cardano.feature/marlowe/src/Language/Marlowe/Scripts/Types.hs 
vimdiff marlowe-plutus.new/marlowe-plutus/src/Language/Marlowe/Plutus/Script.hs \
        marlowe-plutus.main/marlowe-plutus/src/Language/Marlowe/Plutus/Semantics.hs 
```

palas reviewed (changes requested):
Thank you. The diffs where very useful too.
@paluh and I reviewed them, and we have some comments. One he will post himself.
We are not sure whether it is possible to keep `makeIsDataIndexed` with the `asData`, but if it is possible it would be desirable. So that is why I am asking for changes.
In addition to that, I've noticed that `ScriptTypes.hs` corresponds to `Script/Types.hs`, not sure if it would be better to have the same naming in both to make it easier to compare.
In the same line, we could order things in `marlowe-cardano` in the same way than in `marlowe-plutus`. But that would be a different PR, ofc.

palas commented on marlowe-plutus/src/Language/Marlowe/Plutus/Semantics/Types.hs:
```diff
@@ -0,0 +1,523 @@
+      | Assert Observation Contract
+      deriving stock (Generic, Data)
+      deriving newtype (ToData, FromData, UnsafeFromData, Haskell.Eq, Haskell.Ord, Haskell.Show)
+    |]
```
Also for `Action`

bwbush replied:
Using `asData` makes `makeIsDataIndexed` redundant because the patterns include that information. I'll double-check the TH source code for `makeIsDataIndexed`, just to be 100% sure.

bwbush replied:
> In the same line, we could order things in marlowe-cardano in the same way than in marlowe-plutus.

If you're referring to the ordering in the semantics files, we'd have to change those in `marlowe-cardano` because the order in `marlowe-plutus` is (unfortunately) constrained by the way GHC handles TH: i.e., identifiers don't resolve during compilation of the module unless the ordering declares them in a particular sequence.

I think we should reorder and clean up the `marlowe` module in `marlowe-cardano`, maybe when we upgrade to Conway near the start of PI5. Typeclass instances etc. are scattered about the file in a not very logical order. @jhbertra has some ideas for improvements.

palas replied:
Ah, if it is not possible to be explicit in `marlowe-plutus` then it makes sense

paluh replied:
@bwbush isn't `makeIsDataIndexed` replaced in the TS block by `deriving stock (Data,...)` which is "constructor order dependent" derivation of the encoding instance:
* `Action`: https://github.com/input-output-hk/marlowe-plutus/pull/12/files#diff-fcfa0434b7f0a95ec914bd85cc1b722b562d81c2d676c51768a9de4b9e1c0b77R246
* `Contract`: https://github.com/input-output-hk/marlowe-plutus/pull/12/files#diff-fcfa0434b7f0a95ec914bd85cc1b722b562d81c2d676c51768a9de4b9e1c0b77R293

bwbush replied:
Actually, `makeIsDataIndexed` is replaced by `asData`, whose splice expands to explicitly numbering the constructors in their order of appearance. The `deriving stock Data` refers to the type `Data.Data (Data)` in the base library and doesn't particularly relate to Plutus.

bwbush replied:
> Ah, if it is not possible to be explicit in marlowe-plutus then it makes sense

I've added a comment documenting this in 392246d0537e80c9c552956282d1569401f351c0.

paluh reviewed (commented):
@bwbush, we had a somewhat philosophical discussion with @palas that seems pretty important. Let me try to summarize this quickly, along with our understanding of how `asData` works:

* `asData` makes the decoding of the `Contract` lazy (without memoization) - we decode ByteString on demand when we pattern match on a value. This process is repeated on every pattern match.

* If we consider a decoding failure, it will happen lazily as well.

* This is not necessarily a problem because we can argue that the merkleization, which we use widely, encodes a similar kind of laziness.

* Merkleization requires the user to analyze the Contract structure before joining the contract. And we can say that the same is true for non merkleized contracts - user should decode and analyze the contract from the chain (sounds like a truism ;-)).

* The difference with merkleization is that the timeout branch is always revealed (it can contain a contract which may include a `When` with a timeout which has to be revealed as well, and so on. Therefore, the timeout branch must contain a visible and decoded path to a `Close` even in a merkleized contract).

* In the case of `asData`, there's a possibility that a malicious actor puts a datum on the chain containing invalid encoding in the timeout branch.

* In such a case, the guarantee behind the timeout is "in some sens" compromised.

A set of questions arises:

* Is the above understanding of `asData` correct?

* Should we allow this new semantics on the chain?

* Should we perhaps apply `asData` only to the `Case` data type so it preserve similar timeout branch semantics to the merkleized contract?

* If we do this (apply to the `Case` data type only), maybe we can just use `data Contract = ... | When [Case ByteString] Timeout Contract | ...`, which requires full decoding of timeout branch (all the way down to the `Close`) and explicitly says which parts of the contract decoding can crash in the evaluation loop of the semantics? 😉

CC: @jhbertra

bwbush commented:
> Let's consider a hypothetical contract decoded lazily using asData:
> 
> When [ Case (Deposit ...) (When [] timeout INVALID_DATA)] timeout Close
> If the user performs the Deposit the semantics won't enforce decoding of the INVALID_DATA in the same step (we don't have > to pattern match on this case when timeout is not reached).
> If that is true then we end up in a continuation which locks the funds.

The above might not be correct because the Marlowe-Cardano specification requires that that output contract from `computeTransaction` equals the output contract in the `Datum` and we use an `instance Eq Contract` that forces deserialization (i.e., it doesn't compare the serialized bytes but instead compares each field after deserializing them). I need to construct a corrupt datum to that INVALID_DATA would be detected in the first post-creation transaction.

paluh commented:
I started wondering - maybe `asData` generates a validation pass over `BuiltInData` (when performing `fromBuiltinData`) which validates the structure against the schema required by the type?

bwbush commented:
[Here is evidence](https://gist.github.com/bwbush/eb372f077a444021e9ede95699e442b5) that corrupt data is only detected later in the contracts's lifecycle.

bwbush commented:
> In addition to that, I've noticed that ScriptTypes.hs corresponds to Script/Types.hs, not sure if it would be better to have the same naming in both to make it easier to compare.

Fixed in c0b42fcc6dec8b0ca056a8f8e99715779ed5a96e.

bwbush commented:
@palas, @paluh, @ramsay-t in 69decf67ae8304df037827ef685e1898ec57bfbc I removed `asData Contract`, added `asData Case` that defaults to not being used (`--flag=-asdata-case`), and set `asData Action` that defaults to being used (`--flag=+asdata-action`). Thus, the spine on `Contact` is fully deserialized, but deserialization of `Action` is "lazy". This seems the best compromise.

ramsay-t reviewed (approved):
This makes sense to me - we are retaining the option to turn on the faster asData stuff, and the risk seems no worse than with bad Merkelisation. We can revisit this in the New Year if we look at some sort of pre-checking for Marlowe contracts.

paluh commented:
Cool. Thanks!
P.S.
We have to discuss which version we want to expose as `current` next year and probably also plan how we will expose the "unaudited" one in the Runtime.

bwbush commented:
See https://nbviewer.org/gist/bwbush/eb372f077a444021e9ede95699e442b5/PLT-9055.ipynb for test results.
